File: net/JetBrains.FormatRipper/Crypt/SignedMessageVerifier.cs
using System;
using System.Threading.Tasks;
using JetBrains.Annotations;
using JetBrains.SignatureVerifier.BouncyCastle.Cms;
using Org.BouncyCastle.X509.Store;

namespace JetBrains.SignatureVerifier.Crypt
{
  public class SignedMessageVerifier
  {
    private readonly CrlProvider _crlProvider;
    private readonly ILogger _logger;

    public SignedMessageVerifier([CanBeNull] ILogger logger)
      : this(new CrlProvider(logger), logger)
    {
    }

    public SignedMessageVerifier([NotNull] CrlProvider crlProvider, [CanBeNull] ILogger logger)
    {
      _crlProvider = crlProvider ?? throw new ArgumentNullException(nameof(crlProvider));
      _logger = logger ?? NullLogger.Instance;
    }

    public Task<VerifySignatureResult> VerifySignatureAsync(
      [NotNull] SignedMessage signedMessage,
      [NotNull] SignatureVerificationParams signatureVerificationParams)
    {
      if (signedMessage == null)
        throw new ArgumentNullException(nameof(signedMessage));

      if (signatureVerificationParams == null)
        throw new ArgumentNullException(nameof(signatureVerificationParams));

      _logger?.Trace($"Verify with params: {signatureVerificationParams}");

      var certs = signedMessage.SignedData.GetCertificates("Collection");
      var signersStore = signedMessage.SignedData.GetSignerInfos();
      return verifySignatureAsync(signersStore, certs, signatureVerificationParams);
    }

    private async Task<VerifySignatureResult> verifySignatureAsync(
      SignerInformationStore signersStore,
      IX509Store certs,
      SignatureVerificationParams signatureVerificationParams)
    {
      foreach (SignerInformation signer in signersStore.GetSigners())
      {
        var siv = new SignerInfoVerifier(signer, certs, _crlProvider, _logger);
        var result = await siv.VerifyAsync(signatureVerificationParams);

        if (result != VerifySignatureResult.Valid)
          return result;
      }

      return VerifySignatureResult.Valid;
    }
  }
}

File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/crypt/SignedMessageVerifier.kt
package com.jetbrains.signatureverifier.crypt

import com.jetbrains.signatureverifier.ILogger
import com.jetbrains.signatureverifier.NullLogger
import org.bouncycastle.cms.SignerInformationStore
import org.bouncycastle.cert.X509CertificateHolder
import org.bouncycastle.util.Store
import org.jetbrains.annotations.NotNull
import org.jetbrains.annotations.Nullable

open class SignedMessageVerifier {
  private val _crlProvider: CrlProvider
  private val _logger: ILogger

  constructor(@Nullable logger: ILogger) : this(CrlProvider(logger), logger)

  constructor(@NotNull crlProvider: CrlProvider, @Nullable logger: ILogger?) {
    _crlProvider = crlProvider
    _logger = logger ?: NullLogger.Instance
  }

  suspend fun VerifySignatureAsync(
    @NotNull signedMessage: SignedMessage,
    @NotNull signatureVerificationParams: SignatureVerificationParams
  ): VerifySignatureResult {
    _logger.Trace("Verify with params: $signatureVerificationParams")
    val certs = signedMessage.SignedData.certificates
    val signersStore = signedMessage.SignedData.signerInfos
    return verifySignatureAsync(signersStore, certs, signatureVerificationParams)
  }

  private suspend fun verifySignatureAsync(
    signersStore: SignerInformationStore,
    certs: Store<X509CertificateHolder>,
    signatureVerificationParams: SignatureVerificationParams
  ): VerifySignatureResult {
    for (signer in signersStore.signers) {
      val siv = SignerInfoVerifier(signer, certs, _crlProvider, _logger)
      val result = siv.VerifyAsync(signatureVerificationParams)
      if (result != VerifySignatureResult.Valid)
        return result
    }
    return VerifySignatureResult.Valid
  }
}

